hydra-task/src/main/java/com/addthis/hydra/task/source/DataSourceQuery.java

```diff
@@ -0,0 +1,279 @@
+        this.queue = new LinkedBlockingQueue<>(this.buffer);
+    }
+
+    @Override public void init() {
+        localInputSource = Objects.equals(System.getenv("HYDRA_JOBID"), jobId);
+        if (!localInputSource) {
+            throw new IllegalStateException("At this time only queries to ${hydra.task.jobid} are supported");
+        }
+        if (fields != null) {
+            fieldsIn = new IndexField[fields.length];
+            fieldsOut = new AutoField[fields.length];
+            for (int i = 0; i < fields.length; i++) {
+                fieldsIn[i] = new IndexField(i);
+                fieldsOut[i] = AutoField.newAutoField(fields[i]);
+            }
+        }
+        try {
+            File dir = Paths.get(treeDir).toFile();
+            tempDir = Files.createTempDir();
+            queryEngine = new QueryEngineSourceSingle(new ReadTree(dir));
+            Query query = new Query(jobId, new String[] {path}, null);
+            QueryOpProcessor proc = new QueryOpProcessor.Builder(this, new String[]{ops})
+                    .tempDir(tempDir).build();
+            queryHandle = queryEngine.query(query, proc);
+        } catch (Exception ex) {
+            throw Throwables.propagate(ex);
+        }
+    }
+
+    @Override public Bundle next() throws DataChannelError {
+        int countdown = pollCountdown;
+        while ((pollCountdown == 0) || (countdown-- > 0)) {
+            Bundle next = pollAndCloseOnInterrupt(pollInterval, TimeUnit.MILLISECONDS);
+            if (next != null) {
+                return next;
+            }
+            if (closeFuture.isDone()) {
+                closeFuture.join();
+                return null;
+            }
+            if (pollCountdown > 0) {
+                log.info("next polled null, retrying {} more times. shuttingDown={}", countdown, shuttingDown.get());
+            }
+            log.info("null poll ");
+        }
+        if (countdown < 0) {
+            log.info("exit with no data during poll countdown");
+        }
+        return null;
+    }
+
+    private Bundle pollAndCloseOnInterrupt(long pollFor, TimeUnit unit) {
+        boolean interrupted = false;
+        try {
+            long remainingNanos = unit.toNanos(pollFor);
+            long end = System.nanoTime() + remainingNanos;
+            while (true) {
+                try {
+                    return queue.poll(remainingNanos, TimeUnit.NANOSECONDS);
+                } catch (InterruptedException e) {
+                    interrupted = true;
+                    log.info("interrupted while polling for bundles; closing source then resuming poll");
+                    close();
+                    remainingNanos = end - System.nanoTime();
+                }
+            }
+        } finally {
+            if (interrupted) {
+                Thread.currentThread().interrupt();
+            }
+        }
+    }
+
+    @Override public Bundle peek() throws DataChannelError {
+        return queue.peek();
```

Comment:
not allowed to return null from peek before any bundles are produced or you know they never will be